File: backend/modules/petclinic/app/petclinic/api/owners/getOwnerDetail/GetOwnerDetailService.java
package petclinic.api.owners.getOwnerDetail;

import akka.actor.ActorRef;
import akka.pattern.AskSupport;
import akka.pattern.PatternsCS;
import petclinic.model.Owner;
import petclinic.model.OwnerMessages;
import scala.collection.JavaConverters;
import scala.collection.Seq;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;

import static petclinic.api.Constants.TIMEOUT;

public class GetOwnerDetailService implements AskSupport {

    @Inject
    @Named("owner")
    private ActorRef owner;

    public CompletionStage<OwnerGetDetailApiResponse> getOwner(String ownerId) {
        return PatternsCS.ask(owner, OwnerMessages.getOwnerDetails(ownerId), TIMEOUT)
                .thenApply(obj -> (OwnerMessages.GetOwnerDetailsResponse) obj).thenApply(GetOwnerDetailService::mapGetOwnerDetailsResponse);
    }

    static OwnerGetDetailApiResponse mapGetOwnerDetailsResponse(OwnerMessages.GetOwnerDetailsResponse resp) {
        return OwnerGetDetailApiResponse.apply(resp.id(), resp.firstName(), resp.lastName(), resp.address(), resp.city(), resp.telephone(), mapPetsDetailed(resp.pets(), resp.id()));
    }

    static Seq<OwnerGetDetailedPet> mapPetsDetailed(Seq<Owner.Pet> rs, String ownerId) {
        return JavaConverters.asScalaBuffer(JavaConverters.asJavaCollection(rs)
                .stream()
                .map(s -> OwnerGetDetailedPet.apply(s.id(), s.name(), s.birthDay(), s.type(), OwnerGetDetailedPetOwner.apply(ownerId)))
                .collect(Collectors.toList()));
    }
}

File: backend/modules/petclinic/app/petclinic/api/owners/getOwnerDetail/GetOwnerDetailService.scala
package petclinic.api.owners.getOwnerDetail

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.model.owners.owner.Protocol.{GetOwnerDetailsQuery, GetOwnerDetailsResponse}

import scala.concurrent.{ExecutionContext, Future}

class GetOwnerDetailService @Inject() (@Named(petclinic.api.Constants.OWNER_AGGREGATE_ACTOR_NAME) actorRef: ActorRef)(implicit ec: ExecutionContext){

  def getOwner(ownerId: String): Future[OwnerGetDetailApiResponse] = {
    akka.pattern.Patterns.ask(actorRef, GetOwnerDetailsQuery(ownerId), Constants.TIMEOUT)
      .mapTo[GetOwnerDetailsResponse]
      .map(o => OwnerGetDetailApiResponse(o.id, o.firstName, o.lastName, o.address, o.city, o.telephone, o.pets.map(p => {
        OwnerGetDetailedPet(p.id, p.name, p.birthDay, p.`type`, OwnerGetDetailedPetOwner(o.id))
      })))
  }
}
